src/rules/no_constant_condition.rs

```diff
@@ -0,0 +1,443 @@
+// Copyright 2020 the Deno authors. All rights reserved. MIT license.
+use super::Context;
+use super::LintRule;
+use crate::swc_common::Span;
+use crate::swc_common::Spanned;
+use crate::swc_ecma_ast;
+use crate::swc_ecma_ast::Expr;
+use crate::swc_ecma_ast::Lit;
+use crate::swc_ecma_ast::Module;
+use std::sync::Arc;
+use swc_ecma_visit::{Node, Visit};
+
+pub struct NoConstantCondition;
+
+impl LintRule for NoConstantCondition {
+  fn new() -> Box<Self> {
+    Box::new(NoConstantCondition)
+  }
+
+  fn code(&self) -> &'static str {
+    "no-constant-condition"
+  }
+
+  fn lint_module(&self, context: Arc<Context>, module: &Module) {
+    let mut visitor = NoConstantConditionVisitor::new(context);
+    visitor.visit_module(module, module);
+  }
+}
+
+struct NoConstantConditionVisitor {
+  context: Arc<Context>,
+}
+
+impl NoConstantConditionVisitor {
+  pub fn new(context: Arc<Context>) -> Self {
+    Self { context }
+  }
+
+  fn add_diagnostic(&self, span: Span) {
+    self.context.add_diagnostic(
+      span,
+      "no-constant-condition",
+      "Use of a constant expressions as conditions is not allowed.",
+    );
+  }
+
+  fn check_short_circuit(
+    &self,
+    expr: &Expr,
+    operator: swc_ecma_ast::BinaryOp,
+  ) -> bool {
+    match expr {
+      Expr::Lit(lit) => match lit {
+        Lit::Bool(boolean) => {
+          (operator == swc_ecma_ast::BinaryOp::LogicalOr && boolean.value)
+            || (operator == swc_ecma_ast::BinaryOp::LogicalAnd
+              && !boolean.value)
+        }
+        _ => false,
+      },
+      Expr::Unary(unary) => {
+        operator == swc_ecma_ast::BinaryOp::LogicalAnd
+          && unary.op == swc_ecma_ast::UnaryOp::Void
+      }
+      Expr::Bin(bin)
+        if bin.op == swc_ecma_ast::BinaryOp::LogicalAnd
+          || bin.op == swc_ecma_ast::BinaryOp::LogicalOr =>
+      {
+        self.check_short_circuit(&bin.left, bin.op)
+          || self.check_short_circuit(&bin.right, bin.op)
+      }
+      _ => false,
+    }
+  }
+
+  fn is_constant(
+    &self,
+    node: &Expr,
+    parent_node: Option<&Expr>,
+    in_boolean_position: bool,
+  ) -> bool {
+    match node {
+      Expr::Lit(_) | Expr::Arrow(_) | Expr::Fn(_) | Expr::Object(_) => true,
+      Expr::Tpl(tpl) => {
+        (in_boolean_position
+          && tpl.quasis.iter().any(|quasi| match &quasi.cooked {
+            Some(str) => !str.is_empty(),
+            None => false,
+          }))
+          || tpl.exprs.iter().all(|expr| {
+            self.is_constant(expr, parent_node, in_boolean_position)
+          })
+      }
+      // TODO(humancalico) confirm in_boolean_position here
```

Comment:
is this TODO still relevant? What do we need to confirm here?

Reply:
There is no Expression for `Paren` in the original `eslint` implementation and I was not sure what in_boolean_position does (it says `false` if checking branch of a condition. `true` in all other cases) so I am not sure if it'll be `false` here.